File: src/main/java/io/playground/orderservice/infrastructure/persistence/eventstream/OutboxPersistenceAdapter.java
package io.playground.orderservice.infrastructure.persistence.eventstream;

import lombok.RequiredArgsConstructor;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.stereotype.Repository;

import java.time.Instant;
import java.util.List;
import java.util.Map;

@Repository
@RequiredArgsConstructor
public class OutboxPersistenceAdapter {
    private final OutboxJpaRepository outboxRepository;
    private final NamedParameterJdbcTemplate jdbcTemplate;

    public List<OutboxEntity> findNotProcessedOutboxes(int limitSize,
                                                       int retryMax) {
        return outboxRepository.findNotProcessedOutboxes(
                limitSize, retryMax
        );
    }

    public boolean updateProcessedById(Long id, boolean processed) {
        return jdbcTemplate.update(
                "UPDATE outboxes SET " +
                        "processed = :processed " +
                    "WHERE id = :id",
                Map.of(
                        "id", id,
                        "processed", processed
                )
        ) == 1;
    }

    public boolean updateRetryCountAndLockedUntil(Long id, Instant lockedUntil) {
        return jdbcTemplate.update(
                "UPDATE outboxes " +
                        "SET retry_count = retry_count + 1, " +
                        "AND locked_until = :lockedUntil " +
                        "WHERE id = :id",
                Map.of(
                        "id", id,
                        "locked_until", lockedUntil
                )
        ) == 1;
    }
}

File: src/main/kotlin/io/playground/orderservice/infrastructure/persistence/eventstream/OutboxPersistenceAdapter.kt
package io.playground.orderservice.infrastructure.persistence.eventstream

import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.stereotype.Repository
import java.time.Instant

@Repository
class OutboxPersistenceAdapter(
    private val outboxRepository: OutboxJpaRepository,
    private val jdbcTemplate: NamedParameterJdbcTemplate,
) {
    fun findNotProcessedOutboxes(limitSize: Int, retryMax: Int): List<OutboxEntity> =
        outboxRepository.findNotProcessedOutboxes(limitSize, retryMax)

    fun updateProcessedById(id: Long?, processed: Boolean): Boolean =
        jdbcTemplate.update(
            "UPDATE outboxes SET processed = :processed WHERE id = :id",
            mapOf(
                "id" to id,
                "processed" to processed,
            ),
        ) == 1

    fun updateRetryCountAndLockedUntil(id: Long?, lockedUntil: Instant?): Boolean =
        jdbcTemplate.update(
            "UPDATE outboxes SET retry_count = retry_count + 1, AND locked_until = :lockedUntil WHERE id = :id",
            mapOf(
                "id" to id,
                "locked_until" to lockedUntil,
            ),
        ) == 1
}
